core/contextmenu_registry.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Registry for context menu option items.
+ * @author maribethb@google.com (Maribeth Bottorff)
+ */
+'use strict';
+
+/**
+ * @name Blockly.ContextMenuRegistry
+ * @namespace
+ */
+goog.provide('Blockly.ContextMenuRegistry');
+
+goog.require('Blockly.ContextMenuItems');
+
+/**
+ * Class for the registry of context menu items. This is intended to be a singleton. You should
+ * not create a new instance, and only access this class from Blockly.ContextMenuRegistry.registry.
+ * @constructor
+ */
+Blockly.ContextMenuRegistry = function() {
+
+  // Singleton instance should be registered once.
+  Blockly.ContextMenuRegistry.registry = this;
+
+  /**
+   * Registry of all registered RegistryItems, keyed by id.
+   * @type {!Object<string, Blockly.ContextMenuRegistry.RegistryItem>}
+   * @private
+   */
+  this.registry_ = {};
+  Blockly.ContextMenuItems.registerDefaultOptions();
+};
+
+/**
+ * Where this menu item should be rendered. If the menu item should be rendered in multiple
+ * scopes, e.g. on both a block and a workspace, it should be registered for each scope.
+ * @enum {string}
+ * @public
```

Comment:
Implicitly @public, we generally avoid tagging with @public.

Reply:
Done